datafusion/expr/src/type_coercion/functions.rs

```diff
@@ -374,6 +376,67 @@ fn get_valid_types(
             .iter()
             .map(|valid_type| current_types.iter().map(|_| valid_type.clone()).collect())
             .collect(),
+        TypeSignature::String(number) => {
+            if *number < 1 {
+                return plan_err!(
+                    "The signature expected at least one argument but received {}",
+                    current_types.len()
+                );
+            }
+            if *number != current_types.len() {
+                return plan_err!(
+                    "The signature expected {} arguments but received {}",
+                    number,
+                    current_types.len()
+                );
+            }
+
+            fn coercion_rule(
+                lhs_type: &DataType,
+                rhs_type: &DataType,
+            ) -> Result<DataType> {
+                match (lhs_type, rhs_type) {
+                    (DataType::Null, DataType::Null) => Ok(DataType::Utf8),
+                    (DataType::Null, data_type) | (data_type, DataType::Null) => {
+                        coercion_rule(data_type, &DataType::Utf8)
+                    }
+                    (DataType::Dictionary(_, lhs), DataType::Dictionary(_, rhs)) => {
+                        coercion_rule(lhs, rhs)
+                    }
+                    (DataType::Dictionary(_, v), other)
+                    | (other, DataType::Dictionary(_, v)) => coercion_rule(v, other),
+                    _ => {
+                        if let Some(coerced_type) = string_coercion(lhs_type, rhs_type) {
+                            Ok(coerced_type)
+                        } else {
+                            plan_err!(
+                                "{} and {} are not coercible to a common string type",
+                                lhs_type,
+                                rhs_type
+                            )
+                        }
+                    }
+                }
+            }
+
+            // Length checked above, safe to unwrap
+            let mut coerced_type = current_types.first().unwrap().to_owned();
+            for t in current_types.iter().skip(1) {
+                coerced_type = coercion_rule(&coerced_type, t)?;
```

Comment:
This is important decision place. Why would we coerce all arguments to one common super type? Why should a function abstain from accepting a utf8view and utf8 argument pair? fwiw this is exactly what `strpos` was doing?

If we want to do this for some reason, I would prefer the functions to be converted one by one, so that we can eliminate dead code. 

For example if `signature::string(2)` means "any two string-like arguments", then the change in `ContainsFunc` is trivial improvement.
However, if `signature::string(2)` means "two arguments of same string-like type", then  `ContainsFunc` now has dead code which should be removed. (code handling cases of arguments of differing types)


cc @alamb

Reply:
> Why would we coerce all arguments to one common super type? Why should a function abstain from accepting a utf8view and utf8 argument pair? fwiw this is exactly what strpos was doing?

I think this is a simplification for the common case of functions that don't handle all possible String types. Many of the built in function implementations in DataFusion  take multiple string arguments but typically only handle one common type (not all possible combinations of types. 

There is likely a tradeoff in runtime performance with specialized functions for all possible input parameter type combinations and the compiletime / binary size. I think it is reasomable (and common) to coerce to a common super type

Reply:
> this is exactly what strpos was doing?

That is why I think we need function-specific coercion, it is better not to assume the coercion rule for the user, or at least they should be able to define their own coercion rule function-wise. 

Signature::String is a general signature that fit most of the string related function, so we should define the most common way, so it is like what it is now.